File: day2/c++/part1/day2.cpp
/**
 * @file day2.cpp
 * @author ProjectZuki
 * @brief AOC 2024 Day 2 Part 1: Determines safety levels based on defined constraints.
 *          - The levels are either all increasing or all decreasing.
 *          - Any two adjacent levels differ by at least one and at most three.
 * @date 2024-12-02
 */

#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

int main() {

    int safe = 0;
    int unsafe = 0;
    int num;
    // safe if either all increasing or all decreasing
    // safe if any two adjacent levels differ by at least 1 and at most 3

    while (!cin.eof()) {
        vector<int> level;
        while (cin.peek() != '\n' && cin >> num) {
            level.push_back(num);
        }
        cin.ignore();

        // for (int i = 0; i < level.size(); i++) {
        //     cout << level.at(i) << " ";
        // }
        // cout << endl;

        bool isSafe = true;
        bool dec = false;
        bool inc = false;

        for (size_t i = 1; i < level.size() && isSafe; i++) {

            if (level.at(i) > level.at(i-1)) {
                inc = true;
            } else if (level.at(i) < level.at(i-1)) {
                dec = true;
            } else {
                isSafe = false;
                // cout << "one value not increasing or decreasing" << endl;
                break;
            }

            auto diff = abs(level.at(i) - level.at(i-1));
            if (diff > 3 || diff < 1) {
                isSafe = false;
                // cout << "exceeds threshold" << endl;
                
            }

            if (dec && inc) {
                isSafe = false;
                // cout << "increasing and decreasing" << endl;
                break;
            } 
        }

        if (isSafe) {
            safe++;
            // cout << "safe" << endl;
        } else {
            unsafe++;
            // cout << "unsafe" << endl;
        }
        // cout << endl;
    }

    cout << "# safe levels: " << safe << endl;
    cout << "# unsafe levels: " << unsafe << endl;

    return 0;
}

File: day2/python/day2.py

import sys

def main():
    """
    AOC 2024 Day 2 Part 1: Reads a series of integer lists from standard input,
    checks if each list meets the safety conditions, and counts how many lists are safe
    or unsafe based on the following conditions:
    
    - The list is safe if it is either strictly increasing or strictly decreasing.
    - Any two adjacent levels in the list must differ by at least 1 and at most 3.
    
    Outputs the count of safe and unsafe levels.
    """
    safe = 0

    while True:
        # parse until no more input from stdin
        try:
            # Parse single line of input
            line = input().strip()

            # Parse the line into a list of integers
            level = list(map(int, line.split()))

            is_safe = True  # tracks if current level is safe
            dec = False     # tracks if level is decreasing
            inc = False     # tracks if level is increasing

            # check conditions
            for i in range(1, len(level)):
                if level[i] > level[i - 1]:
                    inc = True
                elif level[i] < level[i - 1]:
                    dec = True
                else:
                    is_safe = False
                    break

                diff = abs(level[i] - level[i - 1])
                if diff > 3 or diff < 1:
                    is_safe = False
                    break

                if dec and inc:
                    is_safe = False
                    break

            if is_safe:
                safe += 1

        except EOFError:
            break

    # output safe levels
    print(f"# safe levels: {safe}")

if __name__ == "__main__":
    main()
